README.md
# @vivaxy/reveal

![@vivaxy/reveal](assets/icons/node-reveal.jpg)

🎁 A reveal.js cli

[![NPM Version][npm-version-image]][npm-url]
[![NPM Downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]
[![Standard Version][standard-version-image]][standard-version-url]

![screenshot-1](assets/screenshots/screenshot-1.png)

## Feature

Based on [reveal.js](https://github.com/hakimel/reveal.js#markdown).

- Easy to setup. No cloning repositories. No template files.
- Markdown files as slides content.
- Sync sliding between clients.
- Auto reload when markdown file changes.
- Speaker view.
- Theme config.
- Inline html.
- Support Node.js >= v8.

## Notable Markdown Syntax

- Use `\r?\n----\r?\n` as horizontal slides separator.
- Use `\r?\n---\r?\n` as vertical slides separator.
- Use `^Note:` as speaker notes separator.
- Use `<!-- .slide: data-background="#ff0000" -->` to customize slide styles.
- Use `<!-- .element: style="width: 60%;" -->` to customize element styles.
- Use `<!-- .element: class="fragment" -->` to create fragment.

## Basic Usage

### npx

`npx @vivaxy/reveal server --markdown /path/to/presentation.md`

### Keyboard Shortcuts

- `Space`: Next
- `Up`, `Down`, `Left`, `Right`: Navigation
- `k`, `j`, `h`, `l`: Navigation
- `f`: Full-screen
- `s`: Show slide notes
- `o`: Toggle overview
- `.` (`Period`, `b`, `v` or `/`): Turn screen black
- `Esc`: Escape from full-screen, or toggle overview
- `alt + Left Click`: Toggle zoom
- `p`, `n`: Previous and next

## Advanced Usage

### Commands

#### `server`

##### Basic Usage

```sh
reveal server \
    --markdown ./ppt/reveal.md \
    --theme solarized \
    --highlight-theme solarized-light \
    --transition slide \
    --port 8080 \
    --watch \
    --separator '^\r?\n----\r?\n$' \
    --separator-vertical '^\r?\n---\r?\n$' \
    --separator-notes '^Note:' \
    --width 1440 \
    --height 900 \
    --log-level 2 \
    --script 'custom.js'
```

##### Options

| Name                   | Type          | Required | Default            | Description                                                         |
| ---------------------- | ------------- | -------- | ------------------ | ------------------------------------------------------------------- |
| `--markdown`           | string        | ✔        | N/A                | markdown file                                                       |
| `--theme`              | string        | ✖        | `solarized`        | [`reveal.js` theme](https://revealjs.com/themes/)                   |
| `--highlight-theme`    | string        | ✖        | `solarized-light`  | [`highlight.js` theme](https://highlightjs.org/static/demo/)        |
| `--transition`         | string        | ✖        | `slide`            | [`reveal.js` transitions styles](https://revealjs.com/transitions/) |
| `--port`               | number        | ✖        | `0`                | server port                                                         |
| `--watch`              | boolean       | ✖        | `false`            | reload when markdown changed                                        |
| `--separator`          | string        | ✖        | `^\r?\n----\r?\n$` | horizontal slides separator                                         |
| `--separator-vertical` | string        | ✖        | `^\r?\n---\r?\n$`  | vertical slides separator                                           |
| `--separator-notes`    | string        | ✖        | `^Note:`           | speaker notes separator                                             |
| `--width`              | number        | ✖        | 1440               | slide width                                                         |
| `--height`             | number        | ✖        | 900                | slide height                                                        |
| `--log-level`          | number/string | ✖        | `2`                | [log output level](https://github.com/vivaxy/log-util#log-level)    |
| `--script`             | string        | ✖        | ''                 | insert custom script before head                                    |

### URL Parameters

#### `print-pdf`

Used to print as pdf files.

Add `?print-pdf&showNotes=true`, like: `http://127.0.0.1:3000/?print-pdf#/`.

Press `Ctrl/Command + p` to open print dialog.

Usually, `print-pdf` is used with `showNotes`, like `http://127.0.0.1:3000/?print-pdf&showNotes=separate-page#/`

#### `showNotes`

Show speakers notes.

Add `?showNotes=true`, like: `http://127.0.0.1:3000/?showNotes=true#/`.

## Support

node >= v8.0

## Change Log

See [CHANGELOG.md](CHANGELOG.md).

## Prior Art

- [ksky521/nodePPT](https://github.com/ksky521/nodePPT)
- [hiroppy/fusuma](https://github.com/hiroppy/fusuma)
- [hakimel/reveal.js](https://github.com/hakimel/reveal.js)
- [webpro/reveal-md](https://github.com/webpro/reveal-md)
- [int64ago/node-reveal](https://github.com/int64ago/node-reveal)

[npm-version-image]: http://img.shields.io/npm/v/@vivaxy/reveal.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@vivaxy/reveal
[npm-downloads-image]: https://img.shields.io/npm/dt/@vivaxy/reveal.svg?style=flat-square
[license-image]: https://img.shields.io/npm/l/@vivaxy/reveal.svg?style=flat-square
[license-url]: LICENSE
[standard-version-image]: https://img.shields.io/badge/release-standard%20version-brightgreen.svg?style=flat-square
[standard-version-url]: https://github.com/conventional-changelog/standard-version


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.2.1](https://github.com/vivaxy/node-reveal/compare/v2.2.0...v2.2.1) (2020-07-30)

## [2.2.0](https://github.com/vivaxy/node-reveal/compare/v2.1.5...v2.2.0) (2020-03-07)


### Features

* :arrow_up:  upgrade reveal.js@^3.9.2 ([dc68d0e](https://github.com/vivaxy/node-reveal/commit/dc68d0e))



### [2.1.5](https://github.com/vivaxy/node-reveal/compare/v2.1.4...v2.1.5) (2019-11-17)


### Bug Fixes

* Uncaught SyntaxError: Illegal return statement ([1243026](https://github.com/vivaxy/node-reveal/commit/1243026))



### [2.1.4](https://github.com/vivaxy/node-reveal/compare/v2.1.3...v2.1.4) (2019-11-15)


### Bug Fixes

* :bug: Fix a sync bug when speaker's view is on ([4fe596e](https://github.com/vivaxy/node-reveal/commit/4fe596e))



### [2.1.3](https://github.com/vivaxy/node-reveal/compare/v2.1.2...v2.1.3) (2019-10-29)



### [2.1.2](https://github.com/vivaxy/node-reveal/compare/v2.1.1...v2.1.2) (2019-10-29)



### [2.1.1](https://github.com/vivaxy/node-reveal/compare/v2.1.0...v2.1.1) (2019-05-20)


### Bug Fixes

* :bug:Fix multi devices slide dead lock. ([5833d4f](https://github.com/vivaxy/node-reveal/commit/5833d4f))



## [2.1.0](https://github.com/vivaxy/node-reveal/compare/v2.0.1...v2.1.0) (2019-05-20)


### Features

* :sparkles:Sync sliding between clients. ([724e9de](https://github.com/vivaxy/node-reveal/commit/724e9de))



### [2.0.1](https://github.com/vivaxy/node-reveal/compare/v2.0.0...v2.0.1) (2019-05-20)


### Bug Fixes

* :bug:Fix Cannot find module 'update-notifier' ([a4997c4](https://github.com/vivaxy/node-reveal/commit/a4997c4))



## [2.0.0](https://github.com/vivaxy/node-reveal/compare/v1.7.0...v2.0.0) (2019-05-20)


### Features

* :arrow_up: ([75114b6](https://github.com/vivaxy/node-reveal/commit/75114b6))


### BREAKING CHANGES

* Drop support for node.js <= v8



<a name="1.7.0"></a>
# [1.7.0](https://github.com/vivaxy/node-reveal/compare/v1.6.0...v1.7.0) (2019-05-20)


### Features

* **args:** :sparkles:Support add custom script by --script ([2811e43](https://github.com/vivaxy/node-reveal/commit/2811e43))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/vivaxy/node-reveal/compare/v1.5.0...v1.6.0) (2019-05-20)


### Features

* **pdf-print:** :sparkles:Restrict pdf printing to per-slide per-page ([66b99cb](https://github.com/vivaxy/node-reveal/commit/66b99cb))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/vivaxy/node-reveal/compare/v1.4.3...v1.5.0) (2019-05-20)


### Features

* **style:** :sparkles:Fix slide style, by default to 1440*900. Centering items by flex. ([0027578](https://github.com/vivaxy/node-reveal/commit/0027578))



<a name="1.4.3"></a>
## [1.4.3](https://github.com/vivaxy/node-reveal/compare/v1.4.2...v1.4.3) (2019-05-07)


### Bug Fixes

* :bug:Fix resource error ([48d6929](https://github.com/vivaxy/node-reveal/commit/48d6929))



<a name="1.4.2"></a>
## [1.4.2](https://github.com/vivaxy/node-reveal/compare/v1.4.1...v1.4.2) (2018-10-09)



<a name="1.4.1"></a>
## [1.4.1](https://github.com/vivaxy/node-reveal/compare/v1.4.0...v1.4.1) (2018-10-09)



<a name="1.4.0"></a>
# [1.4.0](https://github.com/vivaxy/node-reveal/compare/v1.3.0...v1.4.0) (2017-06-06)


### Features

* **server:** :sparkles:Toggle `showNotes` from url parameter instead of terminal config ([1407d28](https://github.com/vivaxy/node-reveal/commit/1407d28))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/vivaxy/node-reveal/compare/v1.2.0...v1.3.0) (2017-06-05)


### Features

* **server:** :sparkles:Reopen browser if user inputs anything in terminal ([7080d69](https://github.com/vivaxy/node-reveal/commit/7080d69))
* **server:** :sparkles:Show notes when `watch` is on ([ea87f6f](https://github.com/vivaxy/node-reveal/commit/ea87f6f))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/vivaxy/node-reveal/compare/v1.1.0...v1.2.0) (2017-06-03)


### Features

* **log:** :sparkles:Add logLevel argument ([0336c0e](https://github.com/vivaxy/node-reveal/commit/0336c0e))
* **port:** :sparkles:Support random port ([0a2bc13](https://github.com/vivaxy/node-reveal/commit/0a2bc13))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/vivaxy/node-reveal/compare/v1.0.0...v1.1.0) (2017-06-03)


### Features

* **server:** :sparkles:Remove role controls ([3301ec4](https://github.com/vivaxy/node-reveal/commit/3301ec4))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/vivaxy/node-reveal/compare/v0.5.1...v1.0.0) (2017-06-02)


### Features

* **bin:** :sparkles:Demand command in command line usage. ([8a47117](https://github.com/vivaxy/node-reveal/commit/8a47117))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/vivaxy/node-reveal/compare/v0.5.0...v0.5.1) (2017-06-01)


### Bug Fixes

* **server:** :bug:Fix fragment change not passed to client ([7eac2af](https://github.com/vivaxy/node-reveal/commit/7eac2af))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/vivaxy/node-reveal/compare/v0.4.0...v0.5.0) (2017-06-01)


### Features

* **server:** :sparkles:New Options ([32adc5b](https://github.com/vivaxy/node-reveal/commit/32adc5b))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/vivaxy/node-reveal/compare/v0.3.0...v0.4.0) (2017-06-01)


### Features

* **server:** :sparkles:Add master role controls ([afc8ba9](https://github.com/vivaxy/node-reveal/commit/afc8ba9))
* **server:** :sparkles:Add reload scripts ([d59458c](https://github.com/vivaxy/node-reveal/commit/d59458c))
* **server:** :sparkles:Add title ([68ad45b](https://github.com/vivaxy/node-reveal/commit/68ad45b))
* **server:** :sparkles:Open browser after server started ([8075e76](https://github.com/vivaxy/node-reveal/commit/8075e76))
* **server:** :sparkles:support theme, highlightTheme, transition configurations ([dc80f2f](https://github.com/vivaxy/node-reveal/commit/dc80f2f))

test/demo.js
/**
 * @since 2019-06-18 13:46
 * @author vivaxy
 */
const { handler, builder } = require('../commands/server.js');

(async () => {
  const args = {};
  Object.keys(builder).forEach((key) => {
    if (builder[key].hasOwnProperty('default')) {
      args[key] = builder[key].default;
    }
  });

  await handler({
    ...args,
    markdown: 'test/demo.md',
    theme: 'simple',
    highlightTheme: 'github',
    watch: true,
    logLevel: 0,
    script: 'custom.js',
  });
})();


test/custom.js
/**
 * @since 2019-05-20 02:02
 * @author vivaxy
 */

console.log('Custom script injected!');


commands/server.js
/**
 * @since 2017-05-31 19:42:13
 * @author vivaxy
 */
const nodePath = require('path');

const ip = require('ip');
const ejs = require('ejs');
const Koa = require('koa');
const log = require('log-util');
const fse = require('fs-extra');
const glob = require('glob-promise');
const getPort = require('get-port');
const chokidar = require('chokidar');
const createSocketIo = require('socket.io');
const openBrowser = require('react-dev-utils/openBrowser');

const projectRoot = process.cwd();
const nodeRevealRoot = nodePath.join(__dirname, '..');
const revealJsRoot = nodePath.join(require.resolve('reveal.js'), '..', '..');
const highlightJsRoot = nodePath.join(
  require.resolve('highlight.js'),
  '..',
  '..'
);
const socketIoClientRoot = nodePath.join(
  require.resolve('socket.io-client'),
  '..',
  '..'
);
const templatePath = nodePath.join(nodeRevealRoot, './template/index.ejs');

const socketSet = new Set();

const readTextFile = async (filePath) => {
  return await fse.readFile(filePath, 'utf8');
};

const getResponseType = (filename) => {
  return nodePath.extname(filename);
};

const renderIndexHtml = async ({
  theme,
  highlightTheme,
  transition,
  watch,
  separator,
  separatorVertical,
  separatorNotes,
  width,
  height,
  script,
}) => {
  const template = await readTextFile(templatePath);
  const render = ejs.compile(template);
  return render({
    theme,
    highlightTheme,
    transition,
    watch,
    separator,
    separatorVertical,
    separatorNotes,
    width,
    height,
    script,
  });
};

const responseIndex = async ({
  theme,
  highlightTheme,
  transition,
  watch,
  separator,
  separatorVertical,
  separatorNotes,
  width,
  height,
  script,
}) => {
  return {
    body: await renderIndexHtml({
      theme,
      highlightTheme,
      transition,
      watch,
      separator,
      separatorVertical,
      separatorNotes,
      width,
      height,
      script,
    }),
  };
};

const createKoaSpecificPathMiddleware = ({
  markdown,
  theme,
  highlightTheme,
  transition,
  watch,
  separator,
  separatorVertical,
  separatorNotes,
  width,
  height,
  script,
}) => {
  const responses = {
    '/': responseIndex,
    '/index.html': responseIndex,
    '/node-reveal/reveal.md': async ({ markdown }) => {
      return {
        body: await readTextFile(markdown),
      };
    },
  };

  return async (ctx, next) => {
    const { path } = ctx.request;
    const getResponse = responses[path];
    if (getResponse) {
      const { body } = await getResponse({
        markdown,
        theme,
        highlightTheme,
        transition,
        watch,
        separator,
        separatorVertical,
        separatorNotes,
        width,
        height,
        script,
      });
      ctx.response.body = body;
      ctx.response.type = getResponseType(path);
    } else {
      await next();
    }
  };
};

const createKoaBeginningPathMiddleware = ({ markdown }) => {
  const send = async (ctx, relativePath, root) => {
    ctx.response.body = await fse.readFile(nodePath.join(root, relativePath));
    ctx.response.type = getResponseType(relativePath);
  };

  const responsesForBeginningPath = {
    '/node-reveal': async (ctx, path) => {
      await send(ctx, path, nodeRevealRoot);
    },
    '/reveal.js': async (ctx, path) => {
      await send(ctx, path, revealJsRoot);
    },
    '/highlight.js': async (ctx, path) => {
      await send(ctx, path, highlightJsRoot);
    },
    '/socket.io-client': async (ctx, path) => {
      await send(ctx, path, socketIoClientRoot);
    },
  };

  const markdownRelativePath = nodePath.relative(
    projectRoot,
    nodePath.dirname(markdown)
  );

  return async (ctx) => {
    const { path } = ctx.request;

    const [_1, beginningPath, ...restPath] = path.split('/');
    const getResponseForBeginningPath =
      responsesForBeginningPath[`/${beginningPath}`];
    if (getResponseForBeginningPath) {
      await getResponseForBeginningPath(ctx, `/${restPath.join('/')}`);
    } else {
      // to resolve images and links relative to markdown
      await send(ctx, path, markdownRelativePath);
    }
  };
};

const createServer = ({
  markdown,
  theme,
  highlightTheme,
  transition,
  port,
  watch,
  separator,
  separatorVertical,
  separatorNotes,
  width,
  height,
  script,
}) => {
  const server = new Koa();

  server.use(
    createKoaSpecificPathMiddleware({
      markdown,
      theme,
      highlightTheme,
      transition,
      watch,
      separator,
      separatorVertical,
      separatorNotes,
      width,
      height,
      script,
    })
  );
  server.use(createKoaBeginningPathMiddleware({ markdown }));
  const nativeServer = server.listen(port);
  log.info('[server]', 'started on', port);
  return nativeServer;
};

const createSocket = (server, { markdown }) => {
  const markdownFilename = nodePath.basename(markdown, '.md');

  const io = createSocketIo(server);
  io.on('connection', (socket) => {
    socket.on('disconnect', () => {
      socketSet.delete(socket);
      log.debug('[server]', 'user disconnected', socketSet.size);
    });
    socket.on('slidechanged', syncPageSliding);
    socketSet.add(socket);
    socket.emit('connected', { title: markdownFilename });
    log.debug('[server]', 'user connected', socketSet.size);
  });
};

const startWatch = ({ markdown }) => {
  const watcher = chokidar.watch(markdown);
  watcher.on('change', () => {
    log.debug(
      '[server]',
      'markdown file changed, sending to ' + socketSet.size + ' clients'
    );
    socketSet.forEach((socket) => {
      socket.emit('reload');
    });
  });
};

const syncPageSliding = (event) => {
  log.debug('[server]', 'slide changed ' + JSON.stringify(event));
  socketSet.forEach((socket) => {
    socket.emit('slidechanged', event);
  });
};

const startServer = ({
  markdown,
  theme,
  highlightTheme,
  transition,
  port,
  watch,
  separator,
  separatorVertical,
  separatorNotes,
  width,
  height,
  script,
}) => {
  const server = createServer({
    markdown,
    theme,
    highlightTheme,
    transition,
    port,
    watch,
    separator,
    separatorVertical,
    separatorNotes,
    width,
    height,
    script,
  });
  createSocket(server, { markdown });
  if (watch) {
    startWatch({ markdown });
  }
  const getOpenURL = () => {
    if (watch) {
      return `http://${ip.address()}:${port}/?showNotes=true`;
    }
    return `http://${ip.address()}:${port}/`;
  };
  const openBrowserFunc = () => {
    openBrowser(getOpenURL());
  };
  openBrowserFunc();
  process.stdin.on('data', openBrowserFunc);
};

const getValidThemes = async (matching, ext) => {
  const files = await glob(matching);
  return files.map((file) => {
    return nodePath.basename(file, ext);
  });
};

const argsFormats = {
  async markdown(input) {
    const markdownExists = await fse.pathExists(input);
    if (!markdownExists) {
      log.error('[server]', 'Invalid markdown file');
      process.exit(1);
    }
    return input;
  },
  async theme(input) {
    const validThemes = await getValidThemes(
      nodePath.join(revealJsRoot, 'dist', 'theme', '*.css'),
      '.css'
    );
    if (!validThemes.includes(input)) {
      log.error('[server]', 'Invalid theme.', 'Use:', validThemes.join('/'));
      process.exit(1);
    }
    return input;
  },
  async highlightTheme(input) {
    const validHighlightTheme = await getValidThemes(
      nodePath.join(highlightJsRoot, 'styles', '*.css'),
      '.css'
    );
    if (!validHighlightTheme.includes(input)) {
      log.error(
        '[server]',
        'Invalid highlight theme.',
        'Use:',
        validHighlightTheme.join('/')
      );
      process.exit(1);
    }
    return input;
  },
  async transition(input) {
    const validTransitions = [
      'none',
      'fade',
      'slide',
      'convex',
      'concave',
      'zoom',
    ];
    if (!validTransitions.includes(input)) {
      log.error(
        '[server]',
        'Invalid transition.' + 'Use:',
        validTransitions.join('/')
      );
      process.exit(1);
    }
    return input;
  },
  async port(input) {
    if (!input) {
      return await getPort();
    }
    return input;
  },
};

const parseArgs = async (args) => {
  const formattedArgs = {};
  const formatArgument = async (key) => {
    const format = argsFormats[key];
    const value = args[key];
    if (format) {
      return await format(value);
    }
    return value;
  };
  const formatJobs = Object.keys(args).map(async (key) => {
    return (formattedArgs[key] = await formatArgument(key));
  });
  await Promise.all(formatJobs);
  return formattedArgs;
};

exports.command = 'server';
exports.describe = 'Start a local server';
exports.builder = {
  markdown: {
    demandOption: true,
    describe: 'markdown file',
    type: 'string',
  },
  theme: {
    default: 'solarized',
    type: 'string',
    describe: 'reveal.js theme',
  },
  highlightTheme: {
    default: 'solarized-light',
    type: 'string',
    describe: 'highlight.js theme',
  },
  transition: {
    default: 'slide',
    type: 'string',
    describe: 'reveal.js transition',
  },
  port: {
    default: 0,
    describe: 'server port',
    type: 'number',
  },
  watch: {
    default: false,
    describe: 'watch markdown change',
    type: 'boolean',
  },
  separator: {
    default: '^\r?\n----\r?\n$',
    describe: 'horizontal slides separator',
    type: 'string',
  },
  separatorVertical: {
    default: '^\r?\n---\r?\n$',
    describe: 'vertical slides separator',
    type: 'string',
  },
  separatorNotes: {
    default: '^Note:',
    describe: 'speaker notes separator',
    type: 'string',
  },
  width: {
    default: 1440,
    describe: 'slide width',
    type: 'number',
  },
  height: {
    default: 900,
    describe: 'slide height',
    type: 'number',
  },
  logLevel: {
    default: 2,
    describe: 'log level',
  },
  script: {
    default: '',
    describe: 'insert custom script to head',
  },
};
exports.handler = async (args) => {
  const {
    markdown,
    theme,
    highlightTheme,
    transition,
    port,
    watch,
    separator,
    separatorVertical,
    separatorNotes,
    width,
    height,
    logLevel,
    script,
  } = await parseArgs(args);

  log.setLevel(logLevel);

  startServer({
    markdown,
    theme,
    highlightTheme,
    transition,
    port,
    watch,
    separator,
    separatorVertical,
    separatorNotes,
    width,
    height,
    script,
  });
};
